context-propagation-api/src/test/java/io/micrometer/contextpropagation/ObservationThreadLocalAccessor.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2002-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.contextpropagation;
+
+public class ObservationThreadLocalAccessor implements ThreadLocalAccessor {
+
+    public static final String KEY = "micrometer.observation";
+
+    public static final Namespace OBSERVATION = Namespace.create(KEY);
+
+    private final NamespaceAccessor<ObservationStore> namespaceAccessor = new NamespaceAccessor<>(OBSERVATION);
+
+    @Override
+    public void captureValues(ContextContainer container) {
+        String value = ObservationThreadLocalHolder.holder.get();
+        if (value != null) {
+            namespaceAccessor.putStore(container, new ObservationStore(value));
+        }
+    }
+
+    @Override
+    public void restoreValues(ContextContainer container) {
+        if (namespaceAccessor.isPresent(container)) {
+            ObservationStore store = namespaceAccessor.getStore(container);
+            String observation = store.getObservation();
+            ObservationThreadLocalHolder.holder.set(observation);
+            System.out.println("Observation [" + observation + "]");
+        }
+    }
+
+    @Override
+    public void resetValues(ContextContainer container) {
+        namespaceAccessor.getRequiredStore(container).close();
+    }
```

Comment:
The above `restoreValues()` method checks namespace existence by `namespaceAccessor.isPresent`, but here uses `getRequiredStore()`.
It seems these methods are called from `restoreThreadLocalValues()` and closing the returned `Scope`.

The pre-requisite for `restoreThreadLocalValues()` to properly work is that `captureValues()` needs to create a `Store` in the `namespaceAccessor`.

Per implementation in `ObservationThreadLocalAccessor#captureValues`, `ObservationStore` is only created when `ObservationThreadLocalHolder.holder.get()` is not null.

In the unit test, if I comment out `ObservationThreadLocalHolder.holder.set("hello");`, it fails with followings:
```java
Store must be set
java.lang.IllegalStateException: Store must be set
	at io.micrometer.contextpropagation.NamespaceAccessor.getRequiredStore(NamespaceAccessor.java:78)
	at io.micrometer.contextpropagation.ObservationThreadLocalAccessor.resetValues(ObservationThreadLocalAccessor.java:46)
	at io.micrometer.contextpropagation.SimpleContextContainer.lambda$restoreThreadLocalValues$6(SimpleContextContainer.java:105)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
	at io.micrometer.contextpropagation.SimpleContextContainer.lambda$restoreThreadLocalValues$7(SimpleContextContainer.java:105)
	at io.micrometer.contextpropagation.SimpleThreadLocalAccessorTests.should_capture_thread_local_values(SimpleThreadLocalAccessorTests.java:38)
	...
```

So, I think here also should be guarded like the above method.

```java
    @Override
    public void resetValues(ContextContainer container) {
		if (namespaceAccessor.isPresent(container)) {
			namespaceAccessor.getRequiredStore(container).close();
		}
    }
```

Reply:
Great catch!